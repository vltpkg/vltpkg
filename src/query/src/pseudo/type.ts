import { splitDepID } from '@vltpkg/dep-id/browser'
import type { ParserState } from '../types.ts'
import { asPostcssNodeWithChildren, asTagNode } from '../types.ts'
import { removeDanglingEdges, removeNode } from './helpers.ts'

/**
 * :type(str) Pseudo-Element will match only nodes that are of
 * the same type as the value used. The type is determined by the
 * first part of the dependency ID.
 */
export const type = async (state: ParserState) => {
  const type = asPostcssNodeWithChildren(state.current)
  const selector = asPostcssNodeWithChildren(type.nodes[0])
  const name = asTagNode(selector.nodes[0]).value
  for (const node of state.partial.nodes) {
    const nodeType = splitDepID(node.id)[0]
    if (nodeType !== name) {
      removeNode(state, node)
    }
  }
  removeDanglingEdges(state)
  return state
}
