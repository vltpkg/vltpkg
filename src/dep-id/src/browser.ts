import { error } from '@vltpkg/error-cause'
import { Spec, type SpecOptions } from '@vltpkg/spec/browser'

export const delimiter: Delimiter = '·'
export type Delimiter = '·'

/**
 * Dependency IDs are a URI-encoded set of strings, separated
 * by the {@link Delimiter} character (`'·'`).
 *
 * The first entry is always the specifier type. The rest depend on the
 * type. `git`, `registry`, and `workspace` entries have 3 fields, the rest
 * have 2.
 *
 * - `registry`: `'registry·<registry>·name@specifier'`
 *   The `<registry>` portion can be a known named registry name, or a
 *   url to a registry. If empty, it is the default registry. Examples:
 *   - `··some-package@2.0.1`
 *   - `·npm·whatever@1.2.3`
 *   - `·http%3A%2F%2Fvlt.sh%2F·x@1.2.3`
 * - `git`: `'git·<git remote>·<git selector>'`. For example:
 *   - `git·github:user/project·branchname`
 *   - `git·git%2Bssh%3A%2F%2Fuser%40host%3Aproject.git·semver:1.x`
 * - `workspace`: `'workspace·<path>'`. For example:
 *   - `workspace·src/mything`
 * - `remote`: `'remote·<url>'`
 * - `file`: `'file·<path>'`
 *
 * Lastly, the final portion can contain arbitrary string data, and is
 * used to store peer dep resolutions to maintain the peerDep contract.
 */
export type DepID =
  | `${'' | 'git'}${Delimiter}${string}${Delimiter}${string}${Delimiter}${string}`
  | `${'' | 'git'}${Delimiter}${string}${Delimiter}${string}`
  | `${'file' | 'remote' | 'workspace'}${Delimiter}${string}${Delimiter}${string}`
  | `${'file' | 'remote' | 'workspace'}${Delimiter}${string}`

/**
 * A {@link DepID}, split apart and URI-decoded
 */
export type DepIDTuple =
  | [
      type: 'git',
      gitRemote: string,
      gitSelector: string,
      extra?: string,
    ]
  | [
      type: 'registry',
      registry: string,
      registrySpec: string,
      extra?: string,
    ]
  | [type: 'file', path: string, extra?: string]
  | [type: 'remote', url: string, extra?: string]
  | [type: 'workspace', workspace: string, extra?: string]

const depIDRegExp = new RegExp(
  `^((git)?${delimiter}[^${delimiter}]*${delimiter}[^${delimiter}]*(${
    delimiter
  }[^${delimiter}]*)?$` +
    `|` +
    `^(file|remote|workspace)${delimiter}[^${
      delimiter
    }]*)(${delimiter}[^${delimiter}]*)?$`,
)

export const isDepID = (str: unknown): str is DepID =>
  typeof str === 'string' && depIDRegExp.test(str)

export const asDepID = (str: string): DepID => {
  if (!isDepID(str)) {
    throw error('Expected dep id', {
      found: str,
    })
  }
  return str
}

/**
 * turn a {@link DepIDTuple} into a {@link DepID}
 */
export const joinDepIDTuple = (list: DepIDTuple): DepID => {
  const [type, first, second, extra] = list
  const f = encode(first)
  switch (type) {
    case 'registry':
      return `${delimiter}${f}${delimiter}${encode(second)}${extra ? `${delimiter}${encode(extra)}` : ''}`
    case 'git':
      return `${type}${delimiter}${f}${delimiter}${encode(second)}${extra ? `${delimiter}${encode(extra)}` : ''}`
    default:
      return `${type}${delimiter}${f}${second ? `${delimiter}${encode(second)}` : ''}`
  }
}

// allow @, but otherwise, escape everything urls do
const encode = (s: string): string =>
  encodeURIComponent(s)
    .replaceAll('%40', '@')
    .replaceAll('%2f', '§')
    .replaceAll('%2F', '§')

/**
 * turn a {@link DepID} into a {@link DepIDTuple}
 */
export const splitDepID = (id: string): DepIDTuple => {
  const [type, first = '', second, extra] = id
    .replaceAll('§', '/')
    .split(delimiter, 4)
  const f = decodeURIComponent(first)
  switch (type) {
    case 'git':
    case '': {
      if (second === undefined) {
        throw error(`invalid ${type} id`, { found: id })
      }
      const t: DepIDTuple = [
        type || 'registry',
        f,
        decodeURIComponent(second),
      ]
      if (extra) t.push(decodeURIComponent(extra))
      return t
    }
    case 'file':
    case 'remote':
    case 'workspace': {
      const t: DepIDTuple = [type, f]
      if (second) t.push(decodeURIComponent(second))
      return t
    }
    default: {
      throw error('invalid DepID type', {
        found: type,
        validOptions: ['git', 'file', 'workspace', 'remote', ''],
      })
    }
  }
}

/**
 * Turn a {@link DepID} into a {@link Spec} object
 */
export const hydrate = (
  id: DepID,
  name?: string,
  options: SpecOptions = {},
): Spec => hydrateTuple(splitDepID(id), name, options)

/**
 * Turn a {@link DepIDTuple} into a {@link Spec} object
 */
export const hydrateTuple = (
  tuple: DepIDTuple,
  name?: string,
  options: SpecOptions = {},
) => {
  const [type, first, second] = tuple
  switch (type) {
    case 'remote': {
      if (!first)
        throw error('no remoteURL found on remote id', {
          found: tuple,
        })
      return Spec.parse(name ?? '(unknown)', first)
    }
    case 'file': {
      if (!first) {
        throw error('no file path found on remote id', {
          found: tuple,
        })
      }
      return Spec.parse(name ?? '(unknown)', `file:${first}`, options)
    }
    case 'registry': {
      if (typeof first !== 'string') {
        throw error('no registry url or name in registry ID', {
          found: tuple,
        })
      }
      if (!second) {
        throw error('no name/specifier in registry ID', {
          found: tuple,
        })
      }
      if (!first) {
        // just a normal name@version on the default registry
        const s = Spec.parse(second)
        if (name && s.name !== name) {
          return Spec.parse(`${name}@npm:${second}`)
        } else {
          return s
        }
      }
      if (!/^https?:\/\//.test(first)) {
        const reg = options.registries?.[first]
        if (first !== 'npm' && !reg) {
          throw error('named registry not found in options', {
            name: first,
            found: tuple,
          })
        }
        return Spec.parse(
          name ?? '(unknown)',
          `${first}:${second}`,
          options,
        )
      }
      const s = Spec.parse(
        name ?? '(unknown)',
        `registry:${first}#${second}`,
        options,
      )
      return name && s.final.name !== name ?
          Spec.parse(s.final.name + '@' + s.bareSpec)
        : s
    }
    case 'git': {
      if (!first) {
        throw error('no git remote in git ID', {
          found: tuple,
        })
      }
      return Spec.parse(
        name ?? '(unknown)',
        first + '#' + second,
        options,
      )
    }
    case 'workspace': {
      if (!first) {
        throw error('no name/path on workspace id', { found: tuple })
      }
      return name && name !== first ?
          Spec.parse(name, `workspace:${first}@*`, options)
        : Spec.parse(first, `workspace:*`, options)
    }
  }
}
