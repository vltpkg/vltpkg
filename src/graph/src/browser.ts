import { asDependencyTypeShort } from "./dependencies.js"
import { getBooleanFlagsFromNum } from './lockfile/types.js'

export { asDependencyTypeShort, getBooleanFlagsFromNum }
