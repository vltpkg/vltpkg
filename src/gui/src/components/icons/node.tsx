import { createLucideIcon } from 'lucide-react'

export const Node = createLucideIcon('Node', [
  [
    'path',
    {
      d: 'M11.8751 22C11.5993 22 11.3423 21.9266 11.1032 21.7977L8.65943 20.3461C8.29224 20.1437 8.47584 20.0703 8.58599 20.0336C9.08209 19.8679 9.17428 19.8312 9.68835 19.5375C9.74382 19.5008 9.81725 19.5187 9.87194 19.5562L11.7462 20.6773C11.8196 20.7141 11.9118 20.7141 11.9665 20.6773L19.2978 16.4328C19.3713 16.396 19.408 16.3226 19.408 16.2304V7.76002C19.408 7.66783 19.3713 7.59439 19.2978 7.55767L11.9665 3.33185C11.8931 3.29513 11.8009 3.29513 11.7462 3.33185L4.41486 7.55767C4.34142 7.59439 4.3047 7.68658 4.3047 7.76002V16.2304C4.3047 16.3039 4.34142 16.396 4.41486 16.4328L6.418 17.5906C7.50239 18.1422 8.18208 17.4984 8.18208 16.8554V8.49518C8.18208 8.38502 8.27427 8.27487 8.4024 8.27487H9.33912C9.44928 8.27487 9.55944 8.36705 9.55944 8.49518V16.8554C9.55944 18.307 8.76959 19.1523 7.39145 19.1523C6.96879 19.1523 6.63832 19.1523 5.70081 18.6929L3.77188 17.5906C3.29453 17.3148 3 16.8007 3 16.2492V7.77877C3 7.2272 3.29375 6.71313 3.77188 6.43735L11.1032 2.19278C11.5626 1.93574 12.1876 1.93574 12.647 2.19278L19.9783 6.43735C20.4556 6.71313 20.7502 7.2272 20.7502 7.77877V16.2492C20.7502 16.8007 20.4564 17.3148 19.9783 17.5906L12.647 21.8352C12.4079 21.9445 12.1321 22 11.8751 22ZM14.1353 16.1757C10.9196 16.1757 10.2587 14.7062 10.2587 13.4562C10.2587 13.346 10.3509 13.2359 10.479 13.2359H11.4345C11.5446 13.2359 11.6368 13.3093 11.6368 13.4194C11.7837 14.3929 12.2063 14.871 14.154 14.871C15.6978 14.871 16.3587 14.5218 16.3587 13.6952C16.3587 13.2179 16.1751 12.8687 13.7681 12.6296C11.7649 12.4273 10.5157 11.9866 10.5157 10.3882C10.5157 8.89987 11.7649 8.01783 13.8595 8.01783C16.2111 8.01783 17.3689 8.82643 17.5158 10.5905C17.5158 10.646 17.497 10.7007 17.4603 10.7561C17.4236 10.7929 17.3681 10.8296 17.3134 10.8296H16.3579C16.2658 10.8296 16.1743 10.7561 16.1556 10.664C15.9353 9.653 15.3657 9.32253 13.8587 9.32253C12.1681 9.32253 11.9665 9.91082 11.9665 10.3514C11.9665 10.8843 12.2056 11.0499 14.4837 11.3436C16.7439 11.6374 17.8095 12.0601 17.8095 13.6405C17.7916 15.257 16.4689 16.1757 14.1353 16.1757Z',
      key: '1',
      fill: 'currentColor',
      strokeWidth: '0',
    },
  ],
])

export const NodeWordmark = createLucideIcon('NodeWordmark', [
  [
    'path',
    {
      d: 'M21.8928 12.5235C21.9591 12.4848 22 12.414 22 12.3371V11.8385C22 11.7618 21.9593 11.6911 21.8928 11.6523L20.1212 10.6237C20.0545 10.585 19.9723 10.5848 19.9053 10.6233L18.1227 11.6525C18.0562 11.6911 18.0151 11.762 18.0151 11.8389V13.8969C18.0151 13.9743 18.0566 14.0459 18.124 14.0842L19.8954 15.0935C19.9606 15.131 20.0407 15.1312 20.1066 15.0948L21.1779 14.4991C21.212 14.4801 21.2332 14.4445 21.2334 14.4059C21.2336 14.367 21.2129 14.331 21.1792 14.3117L19.3855 13.2823C19.3519 13.263 19.3313 13.2274 19.3313 13.1891V12.5439C19.3313 12.5053 19.3519 12.4699 19.3851 12.4509L19.9437 12.1289C19.9767 12.1096 20.0178 12.1096 20.0511 12.1289L20.6096 12.4509C20.6431 12.4701 20.6636 12.5055 20.6636 12.5439V13.0514C20.6636 13.0901 20.6842 13.1257 20.7174 13.1448C20.7509 13.1641 20.792 13.1641 20.8255 13.1446L21.8928 12.5235ZM19.9869 12.4269C19.9998 12.4197 20.0155 12.4197 20.0282 12.4269L20.3703 12.6246C20.383 12.632 20.3908 12.6457 20.3908 12.6604V13.0554C20.3908 13.0702 20.383 13.0838 20.3703 13.0912L20.0282 13.2886C20.0155 13.296 19.9996 13.296 19.9869 13.2886L19.645 13.0912C19.6321 13.0838 19.6243 13.0702 19.6243 13.0554V12.6604C19.6243 12.6457 19.6321 12.632 19.645 12.6246L19.9869 12.4269ZM15.3479 13.2488C15.3479 13.2678 15.3377 13.2856 15.321 13.2954L14.7088 13.6483C14.6921 13.658 14.6715 13.658 14.655 13.6483L14.0429 13.2954C14.0261 13.2856 14.016 13.2678 14.016 13.2488V12.5417C14.016 12.5227 14.0264 12.5047 14.0429 12.4949L14.655 12.1416C14.6718 12.1317 14.6923 12.1317 14.709 12.1416L15.3214 12.4949C15.3379 12.5047 15.3481 12.5227 15.3481 12.5417L15.3479 13.2488ZM15.5133 8.01351C15.4801 7.99508 15.4394 7.99551 15.4063 8.01478C15.3735 8.03427 15.3532 8.06943 15.3532 8.10756V10.8432C15.3532 10.8699 15.3388 10.8949 15.3157 10.9084C15.2924 10.9218 15.2636 10.9218 15.2403 10.9084L14.7938 10.6511C14.7273 10.6127 14.6453 10.6127 14.5788 10.6511L12.7957 11.6801C12.729 11.7186 12.6879 11.7898 12.6879 11.8667V13.9253C12.6879 14.0022 12.729 14.0732 12.7955 14.1119L14.5786 15.1416C14.6451 15.1801 14.7273 15.1801 14.794 15.1416L16.5775 14.1119C16.644 14.0732 16.6851 14.0022 16.6851 13.9253V8.79363C16.6851 8.71568 16.6427 8.64367 16.5747 8.60554L15.5133 8.01351ZM9.23801 10.6123C9.30452 10.5737 9.3867 10.5737 9.45322 10.6123L11.2363 11.6411C11.3028 11.6796 11.3439 11.7508 11.3439 11.8277V13.887C11.3439 13.9639 11.3028 14.0351 11.2363 14.0736L9.45343 15.1032C9.38692 15.1416 9.30473 15.1416 9.23801 15.1032L7.45557 14.0736C7.38906 14.0351 7.34796 13.9639 7.34796 13.887V11.8275C7.34796 11.7508 7.38906 11.6796 7.45557 11.6411L9.23801 10.6123ZM5.99551 11.8105C5.99551 11.733 5.95484 11.6619 5.88791 11.6235L4.10546 10.5977C4.07496 10.5799 4.04128 10.5708 4.00718 10.5695H3.98875C3.95486 10.5708 3.92097 10.5799 3.89047 10.5977L2.10782 11.6235C2.04152 11.6619 2 11.733 2 11.8105L2.00381 14.5726C2.00381 14.6112 2.02394 14.647 2.05761 14.6656C2.09066 14.6855 2.13154 14.6855 2.16458 14.6656L3.22367 14.059C3.2906 14.0192 3.33148 13.9491 3.33148 13.8722V12.5818C3.33148 12.5047 3.37215 12.4337 3.43887 12.3952L3.89005 12.1355C3.92351 12.1158 3.96037 12.1065 3.99765 12.1065C4.0345 12.1065 4.07221 12.1158 4.10504 12.1355L4.55579 12.3952C4.62251 12.4337 4.6636 12.5047 4.6636 12.5818V13.8722C4.6636 13.9491 4.70491 14.0196 4.77142 14.059L5.8305 14.6656C5.86376 14.6855 5.90528 14.6855 5.93811 14.6656C5.97094 14.6468 5.99149 14.611 5.99149 14.5726L5.99551 11.8105Z',
      key: '1',
      fill: 'currentColor',
      strokeWidth: '0',
    },
  ],
])
