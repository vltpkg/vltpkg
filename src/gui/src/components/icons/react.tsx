export const React = (props: React.SVGProps<SVGSVGElement>) => {
  return (
    <svg
      width="500"
      height="500"
      viewBox="0 0 500 500"
      fill="none"
      xmlns="http://www.w3.org/2000/svg"
      {...props}>
      <g clip-path="url(#clip0_2967_809)">
        <path
          d="m250.504 284.862c270.046 284.862 285.889 269.029 285.889 249.499c285.889 229.968 270.046 214.136 250.504 214.136c230.961 214.136 215.119 229.968 215.119 249.499c215.119 269.029 230.961 284.862 250.504 284.862z"
          fill="currentColor"
        />
        <path
          d="m250.501 321.96c355.363 321.96 440.371 289.522 440.371 249.509c440.371 209.495 355.363 177.058 250.501 177.058c145.639 177.058 60.6318 209.495 60.6318 249.509c60.6318 289.522 145.639 321.96 250.501 321.96z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.724 285.725c240.155 376.482 310.768 433.837 345.442 413.83c380.116 393.823 365.721 304.031 313.29 213.274c260.859 122.517 190.247 65.1621 155.572 85.1689c120.898 105.176 135.293 194.968 187.724 285.725z"
          stroke="currentColor"
          stroke-width="18.913"
        />
        <path
          d="m187.726 213.277c135.295 304.034 120.9 393.826 155.574 413.833c190.248 433.84 260.861 376.485 313.292 285.728c365.723 194.971 380.118 105.179 345.443 85.172c310.769 65.1652 240.157 122.52 187.726 213.277z"
          stroke="currentColor"
          stroke-width="18.913"
        />
      </g>
      <defs>
        <clipPath id="clip0_2967_809">
          <rect
            width="397"
            height="353"
            fill="currentColor"
            transform="translate(52 73)"
          />
        </clipPath>
      </defs>
    </svg>
  )
}
