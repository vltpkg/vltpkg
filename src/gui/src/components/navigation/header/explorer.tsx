import { cn } from '@/lib/utils.ts'
import { QueryBar } from '@/components/query-bar/index.tsx'
import { RootButton } from '@/components/explorer-grid/root-button.tsx'
import { QueryMatches } from '@/components/explorer-grid/query-matches.tsx'
import { Search, Command } from 'lucide-react'
import { Kbd } from '@/components/ui/kbd.tsx'
import { QueryBuilder } from '@/components/query-builder/index.tsx'
import { useGraphStore } from '@/state/index.ts'
import SaveQueryButton from '@/components/explorer-grid/save-query.tsx'

export const ExplorerHeader = () => {
  const queryInputFocused = useGraphStore(
    state => state.queryInputFocused,
  )
  const updateQueryInputFocused = useGraphStore(
    state => state.updateQueryInputFocused,
  )
  const queryBuilderOpen = useGraphStore(
    state => state.queryBuilderOpen,
  )
  const queryBuilderDisplay = useGraphStore(
    state => state.queryBuilderDisplay,
  )
  const updateQueryBuilderDisplay = useGraphStore(
    state => state.updateQueryBuilderDisplay,
  )

  /**
   * The query builder is put into view when the query input is focused.
   */
  const handleQueryBuilderDisplay = () => {
    updateQueryInputFocused(!queryInputFocused)
    if (!queryInputFocused) {
      updateQueryBuilderDisplay(true)
    }
  }

  return (
    <div className="flex w-full gap-5">
      <RootButton className="rounded-xl" />
      <QueryBar
        onFocus={handleQueryBuilderDisplay}
        onBlur={handleQueryBuilderDisplay}
        tabIndex={0}
        startContent={
          <Search size={20} className="ml-3 text-neutral-500" />
        }
        wrapperClassName={cn(
          queryInputFocused &&
            'relative after:content-[""] after:absolute after:rounded-[14px] after:-inset-1 after:border-[2px] after:border-ring',
          queryBuilderOpen &&
            queryBuilderDisplay &&
            queryInputFocused &&
            'after:border-b-0 after:rounded-b-none',
        )}
        className={cn(
          'relative',
          queryBuilderOpen && queryBuilderDisplay && 'rounded-b-none',
        )}
        endContent={
          <div className="relative mr-3 hidden items-center gap-1 md:flex">
            <QueryBuilder />
            <QueryMatches />
            <SaveQueryButton />
            <Kbd className='before:content-[" "] relative ml-3 before:absolute before:-ml-10 before:h-[0.75rem] before:w-[2px] before:rounded-sm before:bg-neutral-200 before:dark:bg-neutral-700'>
              <Command size={12} />
            </Kbd>
            <Kbd className="text-sm">k</Kbd>
            <div
              className={cn(
                'absolute inset-0 -bottom-2 -right-3 -top-2 z-[-2] rounded-br-xl rounded-tr-xl border-y border-r border-input bg-gradient-to-r from-white/20 to-white backdrop-blur-sm dark:from-muted/5 dark:to-muted/5',
                queryBuilderOpen &&
                  queryBuilderDisplay &&
                  'rounded-br-none',
              )}
            />
          </div>
        }
      />
    </div>
  )
}
