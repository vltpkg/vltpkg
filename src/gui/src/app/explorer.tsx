import { useEffect, type MouseEvent } from 'react'
import { Query } from '@vltpkg/query'
import { SearchBar } from '@/components/search-bar.jsx'
import { Logo } from '@/components/ui/logo.jsx'
import { ExplorerGrid } from '@/components/explorer-grid/index.jsx'
import { useGraphStore } from '@/state/index.js'
import {
  type TransferData,
  type Action,
  type State,
} from '@/state/types.js'
import { load } from '@/state/load-graph.js'
import { Button } from '@/components/ui/button.jsx'
import { LayoutDashboard, Search, Command } from 'lucide-react'
import { ThemeSwitcher } from '@/components/ui/theme-switcher.jsx'
import { Kbd } from '@/components/ui/kbd.jsx'
import { Footer } from '@/components/ui/footer.jsx'

export type ExplorerOptions = {
  projectRoot?: string
}

type StartGraphData = {
  updateHasDashboard: Action['updateHasDashboard']
  updateGraph: Action['updateGraph']
  updateQ: Action['updateQ']
  updateSpecOptions: Action['updateSpecOptions']
  stamp: State['stamp']
}

const startGraphData = async ({
  updateHasDashboard,
  updateGraph,
  updateQ,
  updateSpecOptions,
  stamp,
}: StartGraphData) => {
  const res = await fetch('./graph.json?random=' + stamp)
  const data = (await res.json()) as TransferData & {
    hasDashboard: boolean
  }
  const { graph, specOptions } = load(data)
  const q = new Query({ graph })

  updateHasDashboard(data.hasDashboard)
  updateGraph(graph)
  updateSpecOptions(specOptions)
  updateQ(q)
}

export const Explorer = () => {
  const updateActiveRoute = useGraphStore(
    state => state.updateActiveRoute,
  )
  const updateErrorCause = useGraphStore(
    state => state.updateErrorCause,
  )
  const updateHasDashboard = useGraphStore(
    state => state.updateHasDashboard,
  )
  const updateGraph = useGraphStore(state => state.updateGraph)
  const updateQ = useGraphStore(state => state.updateQ)
  const updateSpecOptions = useGraphStore(
    state => state.updateSpecOptions,
  )
  const stamp = useGraphStore(state => state.stamp)
  // only load graph data when we want to manually update the graph
  // state in the app, to make sure we're controlling it, we use the
  // stamp state as a dependency of `useEffect` to trigger the load.
  useEffect(() => {
    startGraphData({
      updateHasDashboard,
      updateGraph,
      updateQ,
      updateSpecOptions,
      stamp,
    }).catch((err: unknown) => {
      console.error(err)
      updateActiveRoute('/error')
      updateErrorCause('Failed to initialize explorer.')
    })
  }, [stamp])

  return <ExplorerContent />
}

const ExplorerContent = () => {
  const updateActiveRoute = useGraphStore(
    state => state.updateActiveRoute,
  )
  const updateEdges = useGraphStore(state => state.updateEdges)
  const updateNodes = useGraphStore(state => state.updateNodes)
  const hasDashboard = useGraphStore(state => state.hasDashboard)
  const graph = useGraphStore(state => state.graph)
  const query = useGraphStore(state => state.query)
  const q = useGraphStore(state => state.q)

  // updates the query response state anytime the query changes
  // by defining query and q as dependencies of `useEffect` we
  // make sure that this only runs when the query changes
  useEffect(() => {
    async function updateQueryData() {
      if (!q) return
      const queryResponse = await q.search(query)

      updateEdges(queryResponse.edges)
      updateNodes(queryResponse.nodes)

      // make sure we update the URL with the query string
      const state = history.state as
        | undefined
        | { query: string; route: string }
      if (
        !state ||
        query !== state.query ||
        state.route !== '/explore'
      ) {
        history.pushState(
          { query, route: '/explore' },
          '',
          '/explore?query=' + encodeURIComponent(query),
        )
        window.scrollTo(0, 0)
      }
    }
    updateQueryData().catch((err: unknown) => console.error(err))
  }, [query, q])

  const onDashboardButtonClick = (e: MouseEvent) => {
    e.preventDefault()
    updateActiveRoute('/dashboard')
  }

  return (
    <section className="flex grow flex-col justify-between">
      <div>
        <nav
          className="flex gap-4 md:gap-0 px-8 py-4 items-center justify-between border-b-[1px] border-solid"
          role="navigation">
          <div className="flex w-full h-full items-center justify-end">
            <div className="flex items-baseline flex-1">
              <Logo />
              <div className="ml-6">
                <p className="text-md font-medium">Explore</p>
              </div>
            </div>
            <ThemeSwitcher />
            {hasDashboard ?
              <Button
                className="ml-2"
                variant="outline"
                onClick={onDashboardButtonClick}>
                <LayoutDashboard size={12} /> Back to Dashboard
              </Button>
            : ''}
          </div>
        </nav>
        <section className="flex items-center px-8 py-4 border-b-[1px] border-solid">
          <div className="flex flex-col gap-2 w-full">
            {graph?.projectRoot ?
              <p className="text-xs font-mono font-light text-muted-foreground">
                :host-context(file:{graph.projectRoot})
              </p>
            : ''}
            <SearchBar
              tabIndex={0}
              className="w-full bg-muted-foreground/5"
              startContent={
                <Search size={20} className="ml-3" color="#a3a3a3" />
              }
              endContent={
                <div className="hidden md:flex gap-1 mr-3 backdrop-blur-sm">
                  <Kbd>
                    <Command size={12} />
                  </Kbd>
                  <Kbd className="text-sm">k</Kbd>
                </div>
              }
            />
          </div>
        </section>
      </div>
      <ExplorerGrid />
      <Footer />
    </section>
  )
}
