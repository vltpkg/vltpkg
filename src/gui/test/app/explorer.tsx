import {
  test,
  assert,
  expect,
  vi,
  afterAll,
  afterEach,
  beforeAll,
} from 'vitest'
import { setupServer } from 'msw/node'
import { http, HttpResponse } from 'msw'
import type { Query } from '@vltpkg/query'
import type { EdgeLike, GraphLike, NodeLike } from '@vltpkg/graph'
import { cleanup, render } from '@testing-library/react'
import html from 'diffable-html'
import { useGraphStore as useStore } from '@/state/index.js'
import { Explorer } from '@/app/explorer.jsx'
import { joinDepIDTuple } from '@vltpkg/dep-id/browser'

vi.mock('@/components/search-bar.jsx', () => ({
  SearchBar: 'gui-search-bar',
}))

vi.mock('@/components/ui/logo.jsx', () => ({
  Logo: 'gui-logo',
}))

vi.mock('@/components/ui/title.jsx', () => ({
  Title: 'gui-title',
}))

vi.mock('@/components/ui/card.jsx', () => ({
  Card: 'gui-card',
  CardDescription: 'gui-card-description',
  CardHeader: 'gui-card-header',
  CardTitle: 'gui-card-title',
}))

vi.mock('@/components/explorer-grid/index.jsx', () => ({
  ExplorerGrid: 'gui-explorer-grid',
}))

vi.mock('@/components/explorer-grid/setup-project.jsx', () => ({
  SetupProject: 'gui-setup-project',
}))

vi.mock('@/components/explorer-grid/root-button.jsx', () => ({
  RootButton: 'gui-root-button',
}))

export const restHandlers = [
  http.get('/graph.json', () => {
    return HttpResponse.json({
      importers: [
        {
          id: joinDepIDTuple(['file', '.']),
          name: 'root',
          version: '1.0.0',
          mainImporter: true,
          importer: true,
          location: '.',
          dev: false,
          optional: false,
          manifest: { name: 'root', version: '1.0.0' },
        },
      ],
      lockfile: { options: {}, nodes: [], edges: [] },
    })
  }),
]

const server = setupServer(...restHandlers)

expect.addSnapshotSerializer({
  serialize: v => html(v),
  test: () => true,
})

beforeAll(() => server.listen())

afterEach(() => {
  server.resetHandlers()
  const CleanUp = () => (useStore(state => state.reset)(), '')
  render(<CleanUp />)
  cleanup()
})

afterAll(() => server.close())

test('render default', async () => {
  const Container = () => {
    const updateProjectInfo = useStore(
      state => state.updateProjectInfo,
    )
    updateProjectInfo({ tools: ['vlt'], vltInstalled: true })
    return <Explorer />
  }
  render(<Container />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

test('explorer has project root info', async () => {
  const Container = () => {
    const updateProjectInfo = useStore(
      state => state.updateProjectInfo,
    )
    const updateGraph = useStore(state => state.updateGraph)
    updateProjectInfo({ tools: ['vlt'], vltInstalled: true })
    updateGraph({ projectRoot: '/path/to/project' } as GraphLike)
    return <Explorer />
  }
  render(<Container />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})

test('update nodes and edges info on query change', async () => {
  const nodes: NodeLike[] = []
  const edges: EdgeLike[] = []
  const q = {
    search() {
      return { nodes, edges }
    },
  }

  const Container = () => {
    const updateGraph = useStore(state => state.updateGraph)
    const updateProjectInfo = useStore(
      state => state.updateProjectInfo,
    )
    const updateQ = useStore(state => state.updateQ)
    const updateQuery = useStore(state => state.updateQuery)
    updateGraph({ projectRoot: '/path/to/project' } as GraphLike)
    updateProjectInfo({ tools: ['vlt'], vltInstalled: true })
    updateQ(q as unknown as Query)
    updateQuery('#foo')
    return <Explorer />
  }
  render(<Container />)

  let edgesResult
  let nodesResult
  const Retrieve = () => {
    edgesResult = useStore(state => state.edges)
    nodesResult = useStore(state => state.nodes)
    return ''
  }
  render(<Retrieve />)

  assert.deepEqual(
    edgesResult,
    edges,
    'should update edges with result from query',
  )
  assert.deepEqual(
    nodesResult,
    nodes,
    'should update nodes with result from query',
  )
})

test('explorer not vlt installed', async () => {
  const Container = () => {
    const updateGraph = useStore(state => state.updateGraph)
    const updateProjectInfo = useStore(
      state => state.updateProjectInfo,
    )
    updateGraph({ projectRoot: '/path/to/project' } as GraphLike)
    updateProjectInfo({ tools: [], vltInstalled: false })
    return <Explorer />
  }
  render(<Container />)
  expect(window.document.body.innerHTML).toMatchSnapshot()
})
