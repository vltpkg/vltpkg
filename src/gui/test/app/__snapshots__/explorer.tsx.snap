// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`explorer has project root info 1`] = `

<div>
  <section class="flex grow flex-col justify-between bg-white dark:bg-black">
    <div class="flex w-full items-center justify-between border-t-[1px] px-8 pt-4">
      <p class="font-mono text-xs font-light text-muted-foreground">
        :host-context(file:/path/to/project)
      </p>
    </div>
    <section class="flex items-center border-b-[1px] border-solid px-8 py-4">
      <div class="flex w-full flex-row gap-2">
        <gui-root-button>
        </gui-root-button>
        <gui-search-bar
          tabindex="0"
          classname="w-full bg-muted-foreground/5"
          startcontent="[object Object]"
          endcontent="[object Object]"
        >
        </gui-search-bar>
      </div>
    </section>
    <gui-explorer-grid>
    </gui-explorer-grid>
  </section>
</div>

`;

exports[`explorer not vlt installed 1`] = `

<div>
  <gui-setup-project>
  </gui-setup-project>
</div>

`;

exports[`render default 1`] = `

<div>
</div>

`;
